### PR TITLE
Consolidate some Frame APIs

### PR DESCRIPTION
**Pull request by vyasr**
This PR contributes to #9038 by combining common code for `append`, `astype`, `drop`, and `explode`. It's not a small LOC change, but most of the changed lines are just moving code around. Some minor rewrites of `astype` in this PR do significantly speed up those calls (2x faster for <1M rows). Additionally, I've deprecated `append` since it is now deprecated in pandas. Note that it has not been removed in pandas, so we should not remove it until they do.

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -6169,23 +5892,15 @@ def append(
...
-                if (
-                    cols.get_indexer(other._data.to_pandas_index()) >= 0
-                ).all():
-                    other = other.reindex(columns=cols)
-
-        if is_list_like(other):
-            to_concat = [self, *other]
-        else:
-            to_concat = [self, other]
+        elif isinstance(other, list) and not isinstance(other[0], DataFrame):
```
Comment (galipremsagar) [inline python/cudf/cudf/core/dataframe.py:5895]:
```suggestion
        elif isinstance(other, list) and len(other) > 1 and not isinstance(other[0], DataFrame):
```
This check could probably fail if `other` is `[]`

File: python/cudf/cudf/core/series.py [hunk 2]
```diff
@@ -3363,7 +3150,7 @@ def explode(self, ignore_index=False):
 
         Returns
         -------
-        DataFrame
+        Series
```
Comment (galipremsagar) [inline python/cudf/cudf/core/series.py:3153]: oops, nice 👍

File: python/cudf/cudf/core/indexed_frame.py [hunk 3]
```diff
@@ -1778,6 +1798,403 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
...
+            if errors == "raise":
+                raise e
+            elif errors == "warn":
+                import traceback
+
+                tb = traceback.format_exc()
+                warnings.warn(tb)
+            return self
+
+        return self._from_data(data, index=self._index)
```
Comment (galipremsagar) [inline python/cudf/cudf/core/indexed_frame.py:1952]: If `dtype` and `self.dtype` are equal does pandas return same index or do a deep copy? In our previous behavior we were deep copying

File: python/cudf/cudf/core/dataframe.py (hunk 1 above)
Comment (vyasr) [reply python/cudf/cudf/core/dataframe.py:5895]: Good catch.

File: python/cudf/cudf/core/indexed_frame.py (hunk 3 above)
Comment (vyasr) [reply python/cudf/cudf/core/indexed_frame.py:1952]: Actually I'm not quite sure what you're referring to. Is it the `is_dtype_equal` block in the old `Series.astype` implementation?

Comment (vyasr) [reply python/cudf/cudf/core/indexed_frame.py:1952]: Resolved offline: no need to copy since indexes are immutable.

**Comment by vyasr** [comment]
@gpucibot merge

**Pull request merged**